sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManagerSuite.scala

```diff
@@ -302,8 +302,16 @@ class SymmetricHashJoinStateManagerSuite extends StreamTest with BeforeAndAfter
     while (iter.hasNext) iter.next()
   }
 
-  def numRows(implicit manager: SymmetricHashJoinStateManager): Long = {
-    manager.metrics.numKeys
+  def assertNumRows(stateFormatVersion: Int, target: Long)(
+    implicit manager: SymmetricHashJoinStateManager): Unit = {
+    // This suite originally uses HDFSBackStateStoreProvider, which provides instantaneous metrics
+    // for numRows.
+    // But for version 3 with virtual column families, RocksDBStateStoreProvider updates metrics
+    // asynchronously. This means the number of keys obtained from the metrics are very likely
+    // to be outdated right after a put/remove.
+    if (stateFormatVersion <= 2) {
+      assert(manager.metrics.numKeys == target)
+    }
```

Comment:
so this means we are doing no validation for v3. Or this test doesn't run with v3?

Reply:
We skip this check for v3, the test is still run for v3 and it has some other assertions to go through like get/append/update/remove